pkg/operator/controller/ingress/internal_service.go

```diff
@@ -163,14 +164,18 @@ func internalServiceChanged(current, expected *corev1.Service) (bool, *corev1.Se
 		}
 	}
 	// Preserve fields that the API, other controllers, or user may have
-	// modified.
+	// modified.  Note: This list must be kept consistent with
+	// serviceCmpOpts above!
 	updated.Spec.ClusterIP = current.Spec.ClusterIP
+	updated.Spec.ClusterIPs = current.Spec.ClusterIPs
 	updated.Spec.ExternalIPs = current.Spec.ExternalIPs
 	updated.Spec.HealthCheckNodePort = current.Spec.HealthCheckNodePort
+	updated.Spec.IPFamilies = current.Spec.IPFamilies
+	updated.Spec.IPFamilyPolicy = current.Spec.IPFamilyPolicy
 	for i, updatedPort := range updated.Spec.Ports {
 		for _, currentPort := range current.Spec.Ports {
 			if currentPort.Name == updatedPort.Name {
-				updated.Spec.Ports[i].TargetPort = currentPort.TargetPort
+				updated.Spec.Ports[i].NodePort = currentPort.NodePort
```

Comment:
Doesn't this mean that we are allowing the spec.ports.nodePorts to be changed from what we expect, to what is set in current?

Reply:
If something/someone changed the nodePort, and only the nodePort, then L147-148 would return "unchanged" because we ignore it.  However, if something changed the nodePort, and also changed something we don't ignore, and we get to here, then we change it.  Am I missing something?

Reply:
> Doesn't this mean that we are allowing the spec.ports.nodePorts to be changed from what we expect, to what is set in current?

It does.  This is explained in the comments: https://github.com/openshift/cluster-ingress-operator/blob/4a306cb33d26c30a6bda1a2bef92a6c564b3c3b5/pkg/operator/controller/ingress/internal_service.go#L120-L122 https://github.com/openshift/cluster-ingress-operator/blob/4a306cb33d26c30a6bda1a2bef92a6c564b3c3b5/pkg/operator/controller/ingress/internal_service.go#L166-L168

In particular, `NodePort` is set by the API, and so `NodePort` must be ignored in the comparison and preserved during the update so that the operator does not try to revert the value that the API set.  

> If something/someone changed the nodePort, and only the nodePort, then L147-148 would return "unchanged" because we ignore it. However, if something changed the nodePort, and also changed something we don't ignore, and we get to here, then we change it. Am I missing something?

`updated.Spec.Ports[i].NodePort = currentPort.NodePort` copies the value that was observed (`currentPort.NodePort`) to the final version of the service that we will use for the update (`updated`).  That is, the assignment preserves the current `NodePort` value for updates.  

Note that we initialize `updated` by copying the `spec` field from `expected`: https://github.com/openshift/cluster-ingress-operator/blob/4a306cb33d26c30a6bda1a2bef92a6c564b3c3b5/pkg/operator/controller/ingress/internal_service.go#L151-L152

Then if we want to _preserve_ the current value for some subfield, we need to copy the value from `current`: https://github.com/openshift/cluster-ingress-operator/blob/4a306cb33d26c30a6bda1a2bef92a6c564b3c3b5/pkg/operator/controller/ingress/internal_service.go#L169-L178

At the end of all this, `updated` is the result of merging `current` and `expected`, copying the field values that the operator should update from `expected` and copying the field values that the operator should preserve (that is, leave alone) from `current`.

Reply:
Still, 

> If something/someone changed the nodePort, and only the nodePort, then L147-148 would return "unchanged" because we ignore it. 

Does that mean the change to nodePort won't get saved?

Reply:
The `return false, nil` statement on line 147/148 means that `updateInternalService` won't do an update.  In that case, there's no need to preserve `NodePort`'s value—we're not doing an update anyway.

Reply:
Okay.  If we never expect current to hold any values that are different for the ignored fields, then why do we have to preserve them later?

Reply:
The operator preserves the ignored fields so that it doesn't inadvertently update fields that we intend for it to ignore.  

The operator ignores these fields because it isn't supposed to be managing them.  In particular, the API sets `NodePort`, so the operator should never try to update `NodePort`, even if `expected.Spec.Ports[i].NodePort` differs from `current.Spec.Ports[i].NodePort`.  

Maybe the word "expected" is misleading.  `expected` has some fields where the operator manages the field and actually expects the field to have a certain value, as well as some fields where the operator is intended to ignore changes to the field and preserve the current value,, whatever that value may be.  

Ultimately, the operator should update the service if a managed field (for example, `spec.internalTrafficPolicy`) does not have the expected value, and the operator should update that particular field.  However, the operator should not update the service if _only_ ignored fields (for example, `nodePort`) changed, and the operator should not update that field if the operator has to update the service because some managed field value needs an update.